### PR TITLE
Fold safe

### LINKED ISSUE #128: `foldSafe()`
Add a function similar to `logSafe()` that safely calculates fold (i.e., if the denominator is 0)

### PR DESCRIPTION
**Pull request by PavelBal**
Fixes #128

### CODE REVIEW

File: tests/testthat/test-validation-inclusion.R [hunk 1]
```diff
@@ -38,7 +38,14 @@ test_that("isIncluded returns `TRUE` when compound type values are included", {
   expect_true(isIncluded(c("a", "b"), as.factor(c("a", "b"))))
   expect_true(isIncluded(list("a", "b"), as.factor(c("a", "b"))))
   expect_true(isIncluded(as.Date("1970-02-01"), c(as.Date("1970-02-01"), as.Date("1980-12-21"))))
-  expect_true(isIncluded(as.Date("1970-02-01"), list(as.Date("1970-02-01"), as.Date("1980-12-21"))))
+  # Disabling this test as inclusion test does not work with Dat in a list:
```
Comment (PavelBal) [inline tests/testthat/test-validation-inclusion.R:41]: With R4.3 this test somehow fails. See comments in the code - the behavior of `%in%` is strange when `as.Date()` is packed in a list.

File: R/utilities.R [hunk 2]
```diff
@@ -142,3 +142,31 @@ logSafe <- function(x, base = exp(1), epsilon = ospsuiteUtilsEnv$LOG_SAFE_EPSILO
...
+#' inputX <- c(NA, 1, 5, 0, -1)
+#' inputY <- c(1, -1, NA, 0, -1)
+#' folds <- foldSafe(inputX, inputY)
+foldSafe <- function(x, y, epsilon = ospsuiteUtilsEnv$LOG_SAFE_EPSILON) {
+  validateIsSameLength(x, y)
+  x[x <= epsilon] <- epsilon
+  y[y <= epsilon] <- epsilon
+
+  return(x / y)
+}
```
Comment (Yuri05) [inline R/utilities.R:172]: if I try to call a function, it crashes:
`Error: 'ospsuiteUtilsEnv' is not an exported object from 'namespace:ospsuite.utils'`
it works if I declare it like this:
`foldSafe <- function(x, y, epsilon = ospsuite.utils:::ospsuiteUtilsEnv$LOG_SAFE_EPSILON) {`

So either we have to export the environment (not sure about side effects) or add `ospsuite.utils:::` to the argument

**Comment by PavelBal** [comment]
Can you call `logSafe()`? It has exactly the same signature... And somehow this works for me (wonder)

File: tests/testthat/test-utilities.R [hunk 3]
```diff
@@ -52,3 +52,12 @@ test_that("Correct behavior of logSafe", {
   expectedOutput <- c(NA_real_, log10(1), log10(5), log10(defEpsilon), log10(defEpsilon), log10(defEpsilon))
   expect_equal(logSafe(inputVector, base = 10), expectedOutput)
 })
+
+test_that("Correct behavior of foldSafe", {
+  defEpsilon <- getOSPSuiteUtilsSetting("LOG_SAFE_EPSILON")
+  inputX <- c(NA, 1, 5, 0, -1, defEpsilon, 5)
+  inputY <- c(1, -1, NA, 0, -1, defEpsilon, 2)
+
+  expectedOutput <- c(NA_real_, 1 / defEpsilon, NA, defEpsilon / defEpsilon, defEpsilon / defEpsilon, defEpsilon / defEpsilon, 5 / 2)
```
Comment (Yuri05) [inline tests/testthat/test-utilities.R:61]: why is the 1st expected value `NA_real_` but the 3rd just `NA`?

Comment (PavelBal) [reply tests/testthat/test-utilities.R]: Changed the second to `NA_real_` for consistency.

File: R/utilities.R (hunk 2 above)
Comment (Yuri05) [reply R/utilities.R:172]: OK, my bad: shouldn't have just source the new function without installing the whole package.
Assuming it would work.

**Review by Yuri05 — APPROVED** [review]

**Pull request merged**